src/core/MOM_CoriolisAdv.F90

```diff
@@ -1050,6 +1403,499 @@ subroutine gradKE(u, v, h, KE, KEx, KEy, k, OBC, G, GV, US, CS)
+    call weno_seven_reconstruction_2(q8(7:4:-1), c2) ! Reconstruction in the second upwind stencil
+    call weno_seven_reconstruction_3(q8(8:5:-1), c3) ! Reconstruction in the first upwind stencil
+
+    call weno_seven_reconstruction_0(h8(5:2:-1), d0)
+    call weno_seven_reconstruction_1(h8(6:3:-1), d1)
+    call weno_seven_reconstruction_2(h8(7:4:-1), d2)
+    call weno_seven_reconstruction_3(h8(8:5:-1), d3)
+    if (velocity_smoothing) then
+      call weno_seven_weight_0(u8(5:2:-1), b0)    ! Smoothness indicator of the fourth upwind stencil
+      call weno_seven_weight_1(u8(6:3:-1), b1)    ! Smoothness indicator of the third upwind stencil
+      call weno_seven_weight_2(u8(7:4:-1), b2)    ! Smoothness indicator of the second upwind stencil
+      call weno_seven_weight_3(u8(8:5:-1), b3)    ! Smoothness indicator of the first upwind stencil
+    else
+      call weno_seven_weight_0(q8(5:2:-1), b0)
+      call weno_seven_weight_1(q8(6:3:-1), b1)
+      call weno_seven_weight_2(q8(7:4:-1), b2)
+      call weno_seven_weight_3(q8(8:5:-1), b3)
+    endif
+  endif
+
+  tau = abs((b0 - b3) + 3 * (b1 - b2))
+  w0  = C4_35  * fac_fn(tau, b0)
+  w1  = C18_35 * fac_fn(tau, b1)
+  w2  = C12_35 * fac_fn(tau, b2)
+  w3  = C1_35  * fac_fn(tau, b3)
+
+  s = 1. / (w0 + w1 + w2 + w3)
+  w0 = w0 * s   ! Weights of the stencils
+  w1 = w1 * s
+  w2 = w2 * s
+  w3 = w3 * s
+
+  vr = w0 * c0 + w1 * c1 + w2 * c2 + w3 * c3
+  hr = w0 * d0 + w1 * d1 + w2 * d2 + w3 * d3
+
+!  vr = min(max(q4, q5), vr) ; vr = max(min(q4, q5), vr)
+  hr = min(max(h8(4), h8(5)), hr) ; hr = max(min(h8(4), h8(5)), hr) ! Impose a monotonicity limiter
+
+  qr = vr / max(hr, h_tiny)
+
+
+end subroutine weno_seven_h_weight_reconstruction
+
+!> Compute the smoothness indicator for the fourth upwind stencil of the seventh-order WENO scheme
+subroutine weno_seven_weight_0(q4, w0)
+  real, intent(in) :: q4(4)          !< Tracer values on the four-point stencil [A ~> a]
+  real, intent(inout) :: w0          !< Smoothness indicator for this stencil [A2 ~> a2]
+
+  ! Coefficients from Balsara and Shu (2000). The division by 1000 will be normalized out by fac_fn
+  w0 = q4(1) * (2.107 * q4(1) - 9.402 * q4(2) + 7.042 * q4(3) - 1.854 * q4(4)) + &
+     q4(2) * (11.003 * q4(2) - 17.246 * q4(3) + 4.642 * q4(4)) + &
+     q4(3) * (7.043 * q4(3) - 3.882 * q4(4)) + 0.547 * q4(4) * q4(4)
+
+end subroutine weno_seven_weight_0
+
+!> Compute the smoothness indicator for the third upwind stencil of the seventh-order WENO scheme
+subroutine weno_seven_weight_1(q4, w1)
+  real, intent(in) :: q4(4)          !< Tracer values on the four-point stencil [A ~> a]
+  real, intent(inout) :: w1          !< Smoothness indicator for this stencil [A2 ~> a2]
+
+  ! Coefficients from Balsara and Shu (2000). The division by 1000 will be normalized out by fac_fn
+  w1 = q4(1) * (0.547 * q4(1) - 2.522 * q4(2) + 1.922 * q4(3) - 0.494 * q4(4)) + &
```

Comment:
Please add parentheses to this and the other expressions that add 3 or more real values to specify the order of arithmetic so that the answers will be the same for different compilers or compiler settings.